### PR TITLE
Use [[fallthrough]] if VC++ accepts it, regardless of __cplusplus value

### PR DESCRIPTION
**Pull request by lanurmi**
Since VS 2015 Update 3, the C++ standard targeted by the compiler can be
detected through _MSVC_LANG. This macro is like __cplusplus, but reflects the
actual standard understood by the compiler, unlike __cplusplus which by default
is left to an ancient value for compatibility.

### CODE REVIEW

**Comment by vadz** [comment]
Thanks, we can merge this, of course, but this is not the only place where `__cplusplus` is used. Wouldn't it be better to define some `WX_CPLUSPLUS` that would fallback on `_MSVC_LANG` if necessary/possible and use it both here and in any other place where it may be needed?

**Comment by lanurmi** [comment]
> but this is not the only place where `__cplusplus` is used. Wouldn't it be better to define some `WX_CPLUSPLUS` that would fallback on `_MSVC_LANG` if necessary/possible and use it both here and in any other place where it may be needed?

As `_MSVC_LANG` is unavailable before VS2015 (Update 3), I'd consider it mostly useful for checking availability of C++17/C++20.

Searching the whole project for `#if __cplusplus >=` reveals that most checks are about checking `>= 201103L`, followed by `|| wxCHECK_VISUALC_VERSION(<10 or 11 or 14>)`. Different places check for different VC version, probably for a reason, and to maintain backwards compatiblity, those expressions couldn't be really reduced to something simpler.

In other words, sure there could be a `WX_CPLUSPLUS`, but as I see it, using it would only simplify code in the 3 spots that check for either `>= 201703L` or `>= 202002L`.

**Comment by dconnet** [comment]
What about making sure "/Zc:__cplusplus" is passed as a VC compiler option? This will set __cplusplus to the proper value.

**Comment by lanurmi** [comment]
> What about making sure "/Zc:__cplusplus" is passed as a VC compiler option? This will set __cplusplus to the proper value.

What we cannot do is making sure that end users of the library pass that to VC in their own projects. Almost all of the `__cplusplus` checks are in public headers.

> (I have a comment in my code that says this applies to VS2017, not sure what happens in earlier compilers)

Probably a warning about an unknown command line argument, nothing more.

**Comment by lanurmi** [comment]
For reference, here are the two other places where using `_MSVC_LANG` could be used:

```
diff --git a/include/wx/defs.h b/include/wx/defs.h
index 241b899..07a08cd 100644
--- a/include/wx/defs.h
+++ b/include/wx/defs.h
@@ -1303,7 +1303,8 @@ typedef double wxDouble;
     for doing this, that do the same thing as would happen without them, but
     without the warnings.
  */
-#if defined(__cplusplus) && (__cplusplus >= 202002L)
+#if defined(__cplusplus) && (__cplusplus >= 202002L) ||                       \
+    (defined(_MSVC_LANG) && _MSVC_LANG >= 202002L)
     #define wxALLOW_COMBINING_ENUMS_IMPL(en1, en2)                            \
         inline int operator|(en1 v1, en2 v2)                                  \
             { return static_cast<int>(v1) | static_cast<int>(v2); }           \
diff --git a/include/wx/event.h b/include/wx/event.h
index 5e58938..77dbee1 100644
--- a/include/wx/event.h
+++ b/include/wx/event.h
@@ -145,7 +145,7 @@ inline wxEventFunction wxEventFunctionCast(void (wxEvtHandler::*func)(T&))
 // a type of wxEvtHandler method. But with C++17 this doesn't work when the
 // handler is a noexcept function, so we need to cast it to a noexcept function
 // pointer first.
-#if __cplusplus >= 201703L
+#if __cplusplus >= 201703L || (defined(_MSVC_LANG) && _MSVC_LANG >= 201703L)

 namespace wxPrivate
 {
```

**Comment by vadz** [comment]
The other two are required to avoid compilation problems with strict C++17 compilers but don't really bring any benefits, so for now I'll just merge this.
